### ISSUE TITLE
Android: Changes needed to support API 30 (Android 11) requirements from August 2021

### ISSUE THREAD

**Issue by akien-mga** [godotengine/godot#48636]
**Godot version:**
All Godot versions, but realistically only `master`, `3.x` and `3.3`.
Older versions don't support AABs which will become a requirement. (This might finally force EOL for `2.1`, unless someone is really motivated to backport all this work to the `2.1` branch.)

**Issue description:**

Like clockwork, Google Play has updated requirements for new apps and updates that will be enforced from August 2021 (new apps) and November 2021 (updates). These requirements are described in https://developer.android.com/distribute/best-practices/develop/target-sdk ([archive.org link](https://web.archive.org/web/20210501061433/https://developer.android.com/distribute/best-practices/develop/target-sdk)).

From the above documentation, we need to support:

- [x] Publish with the Android App Bundle format. **Supported in 3.3 and later.**
  - [x] AAB should likely be made the default option, and APK support deprecated? I'd keep it in `3.3` but we can possibly remove it in `3.4` (or deprecate and remove in `3.5` once Google Play no longer accepts APKs).
- [x] Use [Play Asset Delivery](https://developer.android.com/guide/playcore/asset-delivery) or [Play Feature Delivery](https://developer.android.com/guide/playcore/feature-delivery) to deliver assets or features that exceed download size of 150MB. Expansion files (OBBs) will no longer be supported for new apps.
  - [x] Where do we stand here for asset delivery features?
- [x] Target API level 30 (Android 11) or above and adjust for [behavioral changes](https://developer.android.com/about/versions/11/behavior-changes-11). Behavioral changes are:
  - [x] Scoped storage enforcement: Access into external storage directories is limited to an app-specific directory and specific types of media that the app has created. #38913
  - [x] Permissions auto-reset: If users haven't interacted with an app for a few months, the system auto-resets the app's sensitive permissions.
  - [x] Background location access: Users must be directed to system settings in order to grant the background location permission to apps.
  - [x] Package visibility: When an app queries for the list of installed apps on the device, the returned list is filtered.

We need to assess all of the above, implement it ASAP for the `3.x` and `master` branches, and cherry-pick to `3.3`. It would be great if we could have this in a `3.3` release at least by late June/July.

@godotengine/android

**Comment by KoBeWi** [comment]
> APK support deprecated? I'd keep it in 3.3 but we can possibly remove it in 3.4

AFAIK only Google Play supports AABs, you can't even install them directly on the device. They are still useful outside the Google Play.

**Comment by m4gr3d** [comment]
> AAB should likely be made the default option, and APK support deprecated? I'd keep it in 3.3 but we can possibly remove it in 3.4 (or deprecate and remove in 3.5 once Google Play no longer accepts APKs).

We would keep apk support (as the non default option) for stores that still relies  on them (e.g: Oculus mobile store).
Also the ability to unpack and repack an apk may be useful in environments where we can't run the regular Android build tools.

> Use Play Asset Delivery or Play Feature Delivery to deliver assets or features that exceed download size of 150MB. Expansion files (OBBs) will no longer be supported for new apps.

As this will be the only available option when `AAB` is the default, we should aim to land these feature sooner rather than later. Here it would be useful to have an example use-case, a game or developer we can work with to ensure the feature behaves as expected and there's no regressions feature wise.
I'll follow up with a potential timeline for this feature.

> Scoped storage enforcement: Access into external storage directories is limited to an app-specific directory and specific types of media that the app has created. #38913

We should aim to land this for Godot 3.4 (which hopefully we can ship prior to August). 

>  -  Permissions auto-reset: If users haven't interacted with an app for a few months, the system auto-resets the app's sensitive permissions.
>  -  Background location access: Users must be directed to system settings in order to grant the background location permission to apps.
>  -  Package visibility: When an app queries for the list of installed apps on the device, the returned list is filtered.

I don't anticipate any changes being required for these.

**Comment by ICatRegister** [comment]
Is progress expected on this issue? 
Got a letter from my publisher : 

> To continue distributing games further to them we would need an update to a target SDK of 30.

**Comment by m4gr3d** [comment]
@ICatRegister yea, we'll be addressing the issue later next week.

**Comment by tolgayaz1991** [comment]
Hi, what is the latest situation about this upgrade? When is the planned time for Godot Engine to support api 30 (Android 11)?

**Comment by lefavilla** [comment]
@m4gr3d have you adressed this issue yet? I stalk this issue since it is created and very concerned about app updates in august.

**Comment by kuruk-mm** [comment]
Any news about this? I want to publish a game of 450 MB size.

**Comment by lefavilla** [comment]
>>[...] **a game or developer we can work with** to ensure the feature behaves as expected and there's no regressions feature wise.
>> I'll follow up with a potential timeline for this feature.

@kuruk-mm m4gr3d has already said, that they look for devs who they can work with. Your app size seems to be an opportunity for this or am I wrong?

**Comment by kuruk-mm** [comment]
> > > [...] **a game or developer we can work with** to ensure the feature behaves as expected and there's no regressions feature wise.
> > > I'll follow up with a potential timeline for this feature.
> 
> @kuruk-mm m4gr3d has already said, that they look for devs who they can work with. Your app size seems to be an opportunity for this or am I wrong?

Thanks! Now I read that :)

> As this will be the only available option when `AAB` is the default, we should aim to land these feature sooner rather than later. Here it would be useful to have an example use-case, a game or developer we can work with to ensure the feature behaves as expected and there's no regressions feature wise.

I offer myself! Let me know. I would help you with anything you need!

**Comment by kyoz** [comment]
I'm using 3.3.2, today i release an app to internal testing and of course google require API 30.

I'v modified `android/build/config.gradle` file like

```
ext.versions = [
    androidGradlePlugin: '4.0.1',
    compileSdk         : 30,
    minSdk             : 18,
    targetSdk          : 30,
    buildTools         : '30.0.3',
    ...
]
```

I'v tried the output apk and it work, aab file is success uploaded to google play console. But is it illegal. @@.

**Comment by lefavilla** [comment]
> 
> 
> I'm using 3.3.2, today i release an app to internal testing and of course google require API 30.
> 
> I'v modified `android/build/config.gradle` file like
> 
> ```
> ext.versions = [
>     androidGradlePlugin: '4.0.1',
>     compileSdk         : 30,
>     minSdk             : 18,
>     targetSdk          : 30,
>     buildTools         : '30.0.3',
>     ...
> ]
> ```
> 
> I'v tried the output apk and it work, aab file is success uploaded to google play console. But is it legal?. @@.

Interesting. Why should it not be legal?

**Comment by ronnywolf42** [comment]
@akien-mga 
hello! 

we wanted to publish our new game today, but we can't upload it to Google Play because of these issues. I don't know how to get help, so i ask here...or is there customer service? It is August now and only one of the tasks on your list have been done. Google annonced these changes for a long time now. what should we do now? we spent months developing the game in godot because we liked the engine and now all is in vain? i am really worried about godot now and our game. should we change engine to Unity finally? I feel like it!

Just increasing the number to 30 will not do. you also have to implement the new requirements, or game will be taken down after. Will this be fixed in next release? And when please?

I am sorry if i'm very confrontational, but this is huge problem for ALL new games which use this engine.

**Comment by Calinou** [comment]
> ..or is there customer service?

Godot has no customer service, and there is no warranty provided as per the [license](https://godotengine.org/license). We also don't offer official commercial support, but nothing prevents third parties from providing commercial support.

> we spent months developing the game in godot because we liked the engine and now all is in vain? i am really worried about godot now and our game. should we change engine to Unity finally? I feel like it!

Don't worry, support for Android API 30 will come eventually :slightly_smiling_face:

**Comment by ronnywolf42** [comment]
I understand that, sorry. But shouldn't fixing this be a top priority for the Godot team? Instead this issue is sitting here since 3 months and almost nothing has been done. what good is an engine or new shiny features when it can't even deploy to a major platform it supposedly supports? :/

**Comment by akien-mga** [comment]
It *is* a priority, but we have limited resources and we do our best with what we have.

Most of the items in the above list are optional and needed only when you actually need those features. What's needed to fulfill these requirements is to target API level 30, which means losing support for the legacy access to external storage which needs to be replaced with scoped storage, and this is not trivial to retrofit in an existing engine. This is being worked on by @m4gr3d.

> Just increasing the number to 30 will not do. you also have to implement the new requirements, or game will be taken down after.

Are you sure about this? My current understanding is that as long as you're not relying on one of the obsoleted features which need to be replaced by new APIs (scoped storage, 150MB assets delivery), targeting API level 30 should work and be a working solution. That's my current plan for 3.3.3 if we can't finalize scoped storage support in coming days.

**Comment by ronnywolf42** [comment]
I know that Godot is OpenSource and that mostly volonteers are working on this. I just tought that critical things like this have greater importance than new features and such.

We use external storage to save progression data, so that it's available when reinstalling it again. If scoped storage can't be done we have to use a server side solution...or just use internal storage.

I just read through the rest of the requirements in the docs. What about changes to compressed resource files (https://developer.android.com/about/versions/11/behavior-changes-11#compressed-resource-file)? Current export contains a resources.arsc file. Is this ok?

Thank you for responding so quickly, and sorry again. I think it will work, i just got too exited.

**Comment by lefavilla** [comment]
> 
> 
> I know that Godot is OpenSource and that mostly volonteers are working on this. I just tought that critical things like this have greater importance than new features and such.

This morning I had a little Idea. Maybe you could try to outsource the delivery part to a external AAR as Godot-Plugin until it gets cold fixed within the engine if needed?

Same thing goes for the media storage manager, what could be a prefered way in your case without depending on a external data server. 

What I saw in the google docs; you could flag the new permissions in the aar lib and then you could manage the data from your user. Within an AAR you could then manage everything and feed it back to your game likewise feed game data to aar. That is just an **Idea** and not tested by me ! Godot and java can handle JSON strings with no problem. So there is no problem to transfer data strings between godot and a aar manager lib.

_Because of my work I have actually not the time to code it by myself._

**Comment by m4gr3d** [comment]
@ronnywolf42 For saving progression data, using internal storage (or [app specific directories](https://developer.android.com/training/data-storage/app-specific)) is indeed the recommended approach. 
This ensures that the system is able to clean all relevant app data when the user decides to uninstall an app.

If you'd like the user to retain game data, we'll be exposing [this configuration](https://developer.android.com/guide/topics/manifest/application-element#fragileuserdata) which will enabled you to do so.

**Comment by akien-mga** [comment]
For those who are currently blocked by us not target SDK 30, I've made a test build here from #50359 which has support for scoped storage:

https://downloads.tuxfamily.org/godotengine/testing/3.3.3-rc.pr50359/

It would be great if you could confirm that it works for you to deploy on device and on Google Play.

Note that it doesn't implement Play Asset Delivery for 150MB+ AABs yet, that will be done later.

**Comment by ronnywolf42** [comment]
@m4gr3d Thank you for working on that and scoped storage.
@akien-mga thank you! i can try your build with our game.

**Comment by sebastianpennino** [comment]
> I'm using 3.3.2, today i release an app to internal testing and of course google require API 30.
> 
> I'v modified `android/build/config.gradle` file like
> 
> ```
> ext.versions = [
>     androidGradlePlugin: '4.0.1',
>     compileSdk         : 30,
>     minSdk             : 18,
>     targetSdk          : 30,
>     buildTools         : '30.0.3',
>     ...
> ]
> ```
> 
> I'v tried the output apk and it work, aab file is success uploaded to google play console. But is it legal?. @@.

For other persons finding this thread, @kyoz is manually modifying the file located on ${your project folder}/android/build/config.gradle, and changing "targetSdk" to 30 (I didn't change de compileSdk and google play didn't complained).

@kyoz: thank you very much, this worked for me (although I consider it a workaround) and allowed me to keep working with the google play console (I'm still months or even more time away from the real release to the public, but this allows me to keep testing the app internally). Thanks.

**Comment by akien-mga** [comment]
> For those who are currently blocked by us not targeting SDK 30, I've made a test build here from #50359 which has support for scoped storage

There's now an official 3.3.3 RC 2 which includes this work: https://godotengine.org/article/release-candidate-godot-3-3-3-rc-2

Please test and confirm that it works to publish new games on Google Play (and if not, what issues seem to remain). I plan to release 3.3.3 stable this week.

**Comment by m4gr3d** [comment]
@kuruk-mm Quick update; we'll try to land support for [Play asset delivery](https://developer.android.com/guide/playcore/asset-delivery) (which enables larger AAB upload) for the `3.4` release, so rough eta would be around late September.
I'll provide more updates once we start testing the feature.

**Comment by m4gr3d** [comment]
@kuruk-mm I've submitted PR #52526 for review which adds Play Asset Delivery support to Godot 3.4.
If you're familiar with building Godot from source, can you give the PR a try for your project.

**Comment by Calinou** [comment]
It looks like level 31+ will be required starting from July: https://godotforums.org/discussion/29262/android-targetsdk-31-or-32-export

**Comment by NanitesNanites** [comment]
> It looks like targeting level 31+ will be required starting from June (or July) 2022: https://godotforums.org/discussion/29262/android-targetsdk-31-or-32-export

According to here: https://support.google.com/googleplay/android-developer/answer/11926878 from August 1st.

**Comment by hani09876** [comment]
> > For those who are currently blocked by us not targeting SDK 30, I've made a test build here from #50359 which has support for scoped storage
> 
> There's now an official 3.3.3 RC 2 which includes this work: https://godotengine.org/article/release-candidate-godot-3-3-3-rc-2
> 
> Please test and confirm that it works to publish new games on Google Play (and if not, what issues seem to remain). I plan to release 3.3.3 stable this week.

It looks like targeting level 31+ will be required starting from June (or July) 2022:
Please do something

**Comment by Zireael07** [comment]
@hani09876: You were asked to stop spamming this request everywhere.

**Comment by filipworksdev** [comment]
Hi guys,. I am testing target SDK level 33 on my build. Starting with Android 12 we need to add 'android:exported="true' in AndroidManifest.xml under <activity> otherwise is not goingto build. I tested the editor built with target 33 and didn't noticed any issues or differences.

**Comment by m4gr3d** [comment]
> Hi guys, I am testing target SDK level 33 on my build. Starting with Android 12 we need to add `android:exported="true"` in AndroidManifest.xml under `<activity>` otherwise is not going to build. I tested the editor built with target 33 and didn't noticed any issues or differences. Seems 30 to 33 might not require any code changes other than that exported issues mentioned above at 33. It may be possible to skip straight to 32 or 33.

@filipworksdev We're updating the target sdk in https://github.com/godotengine/godot/pull/51815.

@akien-mga we'll need to review and merge https://github.com/godotengine/godot/pull/51815 prior to Google's deadline for updating the target sdk.

**Comment by hani09876** [comment]
> > Hi guys, I am testing target SDK level 33 on my build. Starting with Android 12 we need to add `android:exported="true"` in AndroidManifest.xml under `<activity>` otherwise is not going to build. I tested the editor built with target 33 and didn't noticed any issues or differences. Seems 30 to 33 might not require any code changes other than that exported issues mentioned above at 33. It may be possible to skip straight to 32 or 33.
> 
> @filipworksdev We're updating the target sdk in https://github.com/godotengine/godot/pull/51815.
> 
> @akien-mga we'll need to review and merge https://github.com/godotengine/godot/pull/51815 prior to Google's deadline for updating the target sdk.

Is it implemented on Godot 3.5 RC 4

**Comment by Calinou** [comment]
> Is it implemented on Godot 3.5 RC 4

No.

PS: Please don't multi-quote in replies. Remove all quotes except the last reply to prevent threads from looking too cluttered.

**Comment by NanitesNanites** [comment]
@Calinou Is it planned to update the 3.4 branch with this?

**Comment by Calinou** [comment]
> @Calinou Is it planned to update the 3.4 branch with this?

There may be a 3.4 point release for this, but I can't give any guarantees.

**Comment by akien-mga** [comment]
Please test this build which adds full support for Android scoped storage and updated the target API to 32: https://github.com/godotengine/godot/pull/51815#issuecomment-1164504172

If it works well, this should be part of 3.5-stable in coming days.

There should also be a 3.4.5 release that will target API 31 following #62297 - we'll also need confirmation there that it fulfills the August 2022 requirements.

**Comment by akien-mga** [comment]
This is all implemented in 3.5.
